### ISSUE TITLE
Object Store HTTP Rate Limiting

### ISSUE THREAD

**Issue by Kinrany** [apache/arrow-rs-object-store#268]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I'd like to limit the rate of requests I make to S3, to avoid running into throttling on their side.

**Describe the solution you'd like**
There could be a wrapper, similar to `ThrottledStore`, that puts requests in a queue and executes them up to specified throughput.

**Describe alternatives you've considered**
Rate limiting could be baked into ObjectStore implementations that need it, with rules and configuration that apply to them only. This might be better if there is a wide variety of rate limiting rules between different vendors. On the other hand, this is worse if, say, S3-compatible vendors have different rules between them.

**Comment by criccomini** [comment]
@alamb @tustvold We need something like this for SlateDB. Some folks are using `LocalFileSystem` with SlateDB, so rate limiting at the HTTP layer won't work for us. We need to do it at the `object_store` layer (or above). Would you be amenable to a PR that implements a `RateLimitStore` object store that wraps a store (similar to `ThrottleStore`) and implements a token bucket policy? Something like this: https://github.com/slatedb/slatedb/pull/650

**Comment by criccomini** [comment]
@tustvold I took a look at LimitStore. It seems fairly unsophisticated. Some things that would be useful:

1. Per-operation costs
2. Pluggable cost function
3. Pluggable policy

Are you open to adding such support? It would likely be an incompatible API change on the store.

**Comment by tustvold** [comment]
Couple of points

LimitStore is probably the wrong level to limit HTTP stores due to the internal backoff and retry machinery, and should use the HTTP layer

It is unclear to me what the use-case for rate limiting LocalFileSystem is, and therefore what this should look like. 

The linked issue says on startup slatedb makes a huge number of get requests leading to high memory usage, it seems to me that it would be better to instead control the concurrency / memory usage of whatever process is initiating this, rather than constraining the IO. This would be a more reliable fix, whilst also likely more efficient.

Ultimately throttling IO to avoid OOM seems fraught given it is the application that determines how long buffers live - e.g. throttling a `join_all` will just make it slower

**Comment by trungda** [comment]
Besides `LimitStore`, what do we think about something like `AdaptiveStore` that would have [adaptive concurrency control](https://en.wikipedia.org/wiki/Additive_increase/multiplicative_decrease) based on the responses (503/500) from object-store? I have found this approach much more effective in practice rather than having a hardcoded throttling config. It works nicely even when we have multiple clients running in a distributed setup.

**Comment by alamb** [comment]
Maybe it would be a good crate to publish separately (`object_store_acc` )? 

The question is what (other than shifting the maintenance burden) benefit would we get for putting that implementation in object_store crate itself?

I don't think we have a lot of maintenance bandwidth devoted to this crate at the moment to add new features

**Comment by trungda** [comment]
> I don't think we have a lot of maintenance bandwidth devoted to this crate at the moment to add new features

Curious, what are the responsibilities here. My secret desire is to drive more adoption to this crate rather (than aws sdk create :P). I am ok with either way re: publishing object_store_acc crate or merge into the object_store crate similar to `LimitStore`.

**Comment by criccomini** [comment]
> The question is what (other than shifting the maintenance burden) benefit would we get for putting that implementation in object_store crate itself?

I would love to figure this out. As @trungda said, there are a bunch of useful things that a lot of us are having to write over and over again. IMO, it's important for the success of this project to offer such features. This topic is related to https://github.com/apache/arrow-rs-object-store/issues/14.

On the other hand, I'm completely sympathetic to being stretched thin on free+open source work. I, too, am trying to be disciplined about this. But that leaves the project in a funny spot. I might suggest adding a crate to object_store that's got different governance and extra committers, but I'm not at all sure how Apache's model will work with that.

As you say, perhaps the best thing is for a bunch of us to coalesce around some other project+package that provides utilities for object_store. But that gets back to who wants to put in the time to drive it; I'm not sure anyone does.

**Comment by criccomini** [comment]
(Perhaps this is just my pre-AI brain worrying about things that don't matter anymore. LLMs are pretty good at `impl ObjectStore` now. In fact, I just had Codex build me a fully deterministic `LocalFileSystem` for deterministic simulation testing. We also have our own `RetryingObjectStore`, which was vibed. Maybe we should just bin this and let people build what they want in their own repos.)

**Comment by trungda** [comment]
> LLMs are pretty good at impl ObjectStore now.

Man, it's a sad world but I guess this is an industry wide question. I can get LLM implement this pretty fast too.

**Comment by alamb** [comment]
> I would love to figure this out. As @trungda said, there are a bunch of useful things that a lot of us are having to write over and over again. IMO, it's important for the success of this project to offer such features. This topic is related to https://github.com/apache/arrow-rs-object-store/issues/14.

I agree that common patterns (e.g. rancing reads, coalscing requests, local caching, retry policies, deterministic LocalFileSystem, etc) are all valuable and I encourage anyone with the capacity to make a public library and open source it -- I think you would find many users. 

> LLMs are pretty good at impl ObjectStore now.

I like to believe this is also related to our keeping the ObjectStore crate relatively focused and not letting it expand out of control in terms of features and functionality. This crate sets the APIs and enough implementation to get you started. 

Its design also allows easily extending the capabilities when you need something more advanced

> Man, it's a sad world but I guess this is an industry wide question. I can get LLM implement this pretty fast too.

Yes for sure it is an open question. The question is will we need a new model to maintain LLM generated software -- I think well be discovering this over the next year or two

**Comment by alamb** [comment]
> Curious, what are the responsibilities here. My secret desire is to drive more adoption to this crate rather (than aws sdk create :P). I am ok with either way re: publishing object_store_acc crate or merge into the object_store crate similar to LimitStore.

Thank you @trungda  -- I would love help getting more adoption -- what you could do is:
1. Write more about it  (third party blogs / sources tend to have an outsized effect in LLM recommendations)
2. Build that third party ecosystem of integrations / combinators
